0099-Replacing bash as the default shell.md

```diff
@@ -0,0 +1,120 @@
+
+[reference](https://github.com/oilshell/oil/wiki/Migration-Guide)
+
+Some more background context. Originally this is coming from this [PR](https://github.com/NixOS/nixpkgs/pull/105233) by zimbatm. After seeing this, I thought I would implement the changes needed to make the switch and generate a discussion in a PR. Somebody brought to my attention that this change is significant enought that it needs to go through an RFC.
+
+## Examples and Interactions
+
+[examples-and-interactions]: #examples-and-interactions
+
+Some extracts from the post [why use oil](https://www.oilshell.org/why.html)
+
+- better error handling. One of the confusing things about bash is how to handle errors in scripts. `err_exit` was a try at improving that but it has many pitfalls. Oil aims to have more straightforward error handling. Any error will exit, if you do not want that behavior there is the `try my_function` form.
+
+`mkdir /tmp/dest && cp foo /tmp/dest` becomes simply
+
+```Shell
+mkdir /tmp/dest
+cp foo /tmp/dest
+```
+
+- Introduce functions that need to explicitely define their parameters
+
+```Shell
+proc f(first, @rest) { # @ means "the rest of the arguments"
+  write -- $first
+  write -- @rest # @ means "splice this array"
+}
+```
+
+instead of the traditional
+
+```Shell
+f() {
+  echo $1
+  shift 1
+  echo "$@"
+}
+```
+
+those functions are called procs and their variables don't mess with their outer scope.
+
+- Oil has proper associative arrays (dictionaries or hashmaps) that don't have the problems that bash's have.
+
+- No more quotes to prevent splitting
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Currently oilshell has a bus factor of 1. While Andrew has been very motivated and responsive thus far, this could change in the future.
+- Oil shell is pre 1.0 so the syntax might change.
+- Parts of oil are rewritten in c++, while this could be good for performance, new user might discover bugs that are really hard to debug.
+- Oil has had little usage so far, there might be bugs that will just be discovered by having a large user base.
+- Using osh, could be straightforward, but using oil might take more work as basically the existing bash code need to rewritten with `strict_errexit`.
+- The oil documentation is lacking at the moment.
+- Oil has really lofty goals, which I think it's good, but some potential solutions are not documented or not implemented. One example is eggex which are meant to be a replacement for reggexes.
+
+## Alternatives
+
+[alternatives]: #alternatives
+
+- I'm not aware of any other alternative shell that could enable a smooth transition with bash. [Here](https://github.com/oilshell/oil/wiki/Alternative-Shells) is a list of alternative shells.
```

Comment:
I think there are two alternatives to this which should at least be seriously considered:

More provocative maybe than realistic, but POSIX shell could also be an alternative. It would also allow for a transition period where we could keep using bash and make our scripts POSIX compatible (granted, the changes necessary would be far greater than for osh). The two reasons for this would be:

* Nix's env setup (without structuredAttrs) can't make use of any non-standard bash features like arrays. This leads to surprising behavior where setting `*FlagsArray` attributes which `setup.sh` understands will not work as expected.
* Portability: There are far more POSIX shell implementation or POSIX shell compatible shells than bash compatible ones.

While I don't think me or anyone here actually wants to switch to POSIX shell (it's just not worth the effort as there's little to gain), I do think that answering this question could help us and the RFC to clarify *what* we want from `stdenv.shell`.

The **second** alternative would be to rely less on the shell in the future. The bulk of shell scripts subject to this RFC or discussion is not the build scripts which are quite simple execute-commands-in-sequence affairs, but the setup hooks and the many wrappers (binutils, cc, pkg-config, ...) contained in nixpkgs. [setup hooks have always been considered an implementation detail of nixpkgs](https://nixos.org/manual/nixpkgs/stable/#ssec-setup-hooks) and consequently, we could consider replacing these shell scripts with an alternative approach, e. g.  a mechanism implemented in pure Nix or using a [DSL for manipulating the environment](https://shopify.github.io/shadowenv/) (which is the main thing the wrappers / setup hooks are doing after all).

Should we want to take such a direction in the future, it may not be considered worth it to invest time into Oil support in the short / medium term.